File: android/src/main/java/com/skjline/rideraid/ui/fragment/Telemetry.java
package com.skjline.rideraid.ui.fragment;

import android.annotation.SuppressLint;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.v4.app.Fragment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.skjline.rideraid.R;
import com.skjline.rideraid.RiderAidApplication;
import com.skjline.rideraid.ant.AntBikeDevice;
import com.skjline.rideraid.presenter.PositionPresenter;
import com.skjline.rideraid.presenter.PresenterActions;
import com.skjline.rideraid.presenter.TelemetryPresenter;
import com.skjline.rideraid.presenter.TimePresenter;
import com.skjline.rideraid.presenter.module.Position;
import com.skjline.rideraid.presenter.module.TeleSensor;
import com.skjline.rideraid.presenter.module.Tick;
import com.skjline.rideraid.util.UnitUtils;
import com.skjline.rideraid.util.WindowsUtils;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.TimeZone;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;

public class Telemetry extends Fragment
        implements TimePresenter.View, TelemetryPresenter.View, PositionPresenter.View {
    @BindView(R.id.tv_time_value)
    TextView tvTimer;
    @BindView(R.id.tv_cad_value)
    TextView tvCadence;
    @BindView(R.id.tv_spd_value)
    TextView tvSpeed;
    @BindView(R.id.tv_dist_value)
    TextView tvDistance;

    @BindView(R.id.tv_dist_unit)
    TextView tvDistanceUnit;
    @BindView(R.id.tv_spd_unit)
    TextView tvSpeedUnit;

    Unbinder unbinder;

    private PresenterActions[] presenters;
    private SimpleDateFormat formatter;

    AntBikeDevice device = (AntBikeDevice) RiderAidApplication.ant;

    private int unit;

    @SuppressLint("SimpleDateFormat")
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // Suppressing simple data format since the locale doesn't matter
        formatter = new SimpleDateFormat("HH:mm:ss");
        formatter.setTimeZone(TimeZone.getTimeZone("GMT"));

        presenters = new PresenterActions[]{
                new Tick(this),
                new TeleSensor(this, device),
                new Position(getContext(), this)};

        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(getContext());

        Log.d(getClass().getSimpleName(), "Register shared preference listener");
        pref.registerOnSharedPreferenceChangeListener(listener);

        String str = pref.getString(getString(R.string.unit_type_key), getString(R.string.unit_type_default));
        unit = str.equals(getString(R.string.unit_speed_mph)) ? R.string.unit_speed_mph : R.string.unit_speed_kmh;

        setHasOptionsMenu(true);
        setRetainInstance(true);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.telemetry_fragment, container, false);

        unbinder = ButterKnife.bind(this, view);
        setDisplayUnits(unit);

        return view;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();

        if (unbinder != null) {
            unbinder.unbind();
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        Log.d(getClass().getSimpleName(), "Unregister shared preference listener");
        PreferenceManager.getDefaultSharedPreferences(getContext())
                .unregisterOnSharedPreferenceChangeListener(listener);
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);

        inflater.inflate(R.menu.menu_main, menu);

        PresenterActions p = presenters[0];
        menu.findItem(R.id.action_record).setVisible(!p.isActive());
        menu.findItem(R.id.action_stop).setVisible(p.isActive());
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_record:
                for (PresenterActions p : presenters) {
                    p.start();
                }

                WindowsUtils.lockScreenDim(getActivity().getWindow(), true);
                getActivity().invalidateOptionsMenu();
                return true;
            case R.id.action_stop:
                for (PresenterActions p : presenters) {
                    p.stop();
                }

                WindowsUtils.lockScreenDim(getActivity().getWindow(), false);
                getActivity().invalidateOptionsMenu();
                return true;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    public void updateTime(long time) {
        tvTimer.setText(formatter.format(new Date(time * 1000)));
    }

    @Override
    public void updateCadence(long cad) {
        tvCadence.setText(String.valueOf(cad));
    }

    @Override
    public void updateSpeed(long spd) {
        tvSpeed.setText(String.valueOf(UnitUtils.convertCalcSpeed(spd, unit)));
    }

    @Override
    public void updateDistance(double distance) {
        tvDistance.setText(String.valueOf(UnitUtils.convertDistance(distance, unit)));
    }

    private void setDisplayUnits(int selection) {
        if (selection == R.string.unit_speed_mph) {
            tvDistanceUnit.setText(getString(R.string.unit_mile));
            tvSpeedUnit.setText(getString(R.string.unit_speed_mph));
            return;
        }

        tvDistanceUnit.setText(getString(R.string.unit_km));
        tvSpeedUnit.setText(getString(R.string.unit_speed_kmh));
    }

    SharedPreferences.OnSharedPreferenceChangeListener listener = (sharedPreferences, key) -> {
        if (key.equals(getString(R.string.unit_type_key))) {
            String str = sharedPreferences.getString(getString(R.string.unit_type_key), getString(R.string.unit_type_default));
            unit = str.equals(getString(R.string.unit_speed_mph)) ? R.string.unit_speed_mph : R.string.unit_speed_kmh;
            setDisplayUnits(unit);
        }
    };
}

File: rider/src/main/java/com/pss9/rider/ride/Telemetry.kt
package com.pss9.rider.ride

import android.os.Bundle
import android.preference.PreferenceManager
import android.view.*
import android.widget.TextView
import androidx.fragment.app.Fragment
import com.pss9.rider.R
import com.pss9.rider.RiderAidApplication
import com.pss9.rider.ant.AntBikeDevice
import com.pss9.rider.presenter.PositionPresenter
import com.pss9.rider.presenter.PresenterService
import com.pss9.rider.presenter.TelemetryPresenter
import com.pss9.rider.presenter.TimePresenter
import com.pss9.rider.ride.module.Position
import com.pss9.rider.ride.module.TeleSensor
import com.pss9.rider.ride.module.Tick
import com.pss9.rider.util.UnitUtils
import com.pss9.rider.util.WindowsUtils
import java.text.SimpleDateFormat
import java.util.*

class Telemetry : Fragment(), TimePresenter.View, TelemetryPresenter.View, PositionPresenter.View {
    private lateinit var tvTimer: TextView
    private lateinit var tvCadence: TextView
    private lateinit var tvSpeed: TextView
    private lateinit var tvDistance: TextView

    private lateinit var tvDistanceUnit: TextView
    private lateinit var tvSpeedUnit: TextView

    private var presenters = mutableListOf<PresenterService>()
    private var formatter: SimpleDateFormat =
        SimpleDateFormat("HH:mm:ss", Locale.getDefault()).apply {
            timeZone = TimeZone.getTimeZone("GMT")
        }

    private val device: AntBikeDevice = RiderAidApplication.ant

    private var unit = 0
    private val pref by lazy { PreferenceManager.getDefaultSharedPreferences(context) }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        presenters.addAll(listOf(Tick(this), TeleSensor(this, device), Position(context!!, this)))

        setHasOptionsMenu(true)
        retainInstance = true
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View = inflater.inflate(R.layout.telemetry_fragment, container, false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        tvTimer = view.findViewById(R.id.tv_time_value)
        tvCadence = view.findViewById(R.id.tv_cad_value)
        tvSpeed = view.findViewById(R.id.tv_spd_value)
        tvDistance = view.findViewById(R.id.tv_dist_value)

        tvDistanceUnit = view.findViewById(R.id.tv_dist_unit)
        tvSpeedUnit = view.findViewById(R.id.tv_spd_unit)
    }

    override fun onResume() {
        super.onResume()
        setDisplayUnits()
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)

        inflater.inflate(R.menu.menu_main, menu)

        val p = presenters.first()

        menu.findItem(R.id.action_record).isVisible = !p.isActive()
        menu.findItem(R.id.action_stop).isVisible = p.isActive()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_record, R.id.action_stop -> {
                for (p in presenters) {
                    p.start()
                }

                WindowsUtils.lockScreenDim(activity!!.window, item.itemId == R.id.action_record)
                activity!!.invalidateOptionsMenu()
                return true
            }
        }

        return super.onOptionsItemSelected(item)
    }

    override fun updateTime(second: Long) {
        tvTimer.text = formatter.format(Date(second * 1000))
    }

    override fun updateCadence(cad: Long) {
        tvCadence.text = cad.toString()
    }

    override fun updateSpeed(spd: Long) {
        tvSpeed.text = UnitUtils.convertCalcSpeed(spd, unit).toString()
    }

    override fun updateDistance(distance: Double) {
        tvDistance.text = UnitUtils.convertDistance(distance, unit).toString()
    }

    private fun setDisplayUnits() {

        val str = pref.getString(getString(R.string.unit_type_key), getString(R.string.unit_type_default))
        unit = if (str == getString(R.string.unit_speed_mph)) {
            R.string.unit_speed_mph
        } else {
            R.string.unit_speed_kmh
        }

        if (unit == R.string.unit_speed_mph) {
            tvDistanceUnit.text = getString(R.string.unit_mile)
            tvSpeedUnit.text = getString(R.string.unit_speed_mph)
        } else {
            tvDistanceUnit.text = getString(R.string.unit_km)
            tvSpeedUnit.text = getString(R.string.unit_speed_kmh)
        }
    }
}
